src/crypto/shelley/ledgerUtils.js

```diff
@@ -0,0 +1,905 @@
+  try {
+    Logger.debug('ledgerUtils::getHWDeviceInfo called')
+
+    const transport = await connectionHandler(deviceId, deviceObj, useUSB)
+    const appAda = new AppAda(transport)
+    const versionResp: GetVersionResponse = await appAda.getVersion()
+    Logger.debug('AppAda version', versionResp)
+
+    // assume single account in Yoroi
+    const accountPath = makeCardanoAccountBIP44Path(
+      getWalletType(walletImplementationId),
+      CONFIG.NUMBERS.ACCOUNT_INDEX,
+    )
+    Logger.debug('bip44 account path', accountPath)
+
+    // get Cardano's first account
+    // i.e hdPath = [2147483692, 2147485463, 2147483648]
+    const extendedPublicKeyResp: GetExtendedPublicKeyResponse = await appAda.getExtendedPublicKey(
+      accountPath,
+    )
+    Logger.debug('extended public key', extendedPublicKeyResp)
+    Logger.debug('transport.id', transport.id)
+
+    const hwDeviceInfo = normalizeHWResponse({
+      extendedPublicKeyResp,
+      deviceId,
+      deviceObj,
+    })
+    Logger.info('ledgerUtils::getHWDeviceInfo: Ledger device OK')
+    Logger.info('hwDeviceInfo', hwDeviceInfo)
+    await transport.close()
+    return hwDeviceInfo
+  } catch (e) {
+    if (_isUserError(e)) {
+      Logger.info('ledgerUtils::getHWDeviceInfo: User-side error', e)
+      throw new LedgerUserError()
+    } else if (_isConnectionError(e)) {
+      Logger.info('ledgerUtils::getHWDeviceInfo: General/BleError', e)
+      throw new GeneralConnectionError()
+    } else {
+      Logger.error('ledgerUtils::getHWDeviceInfo: Unexpected error', e)
+      throw e
+    }
+  }
+}
+
+export const verifyAddress = async (
+  walletImplementationId: WalletImplementationId,
+  address: string,
+  addressing: $PropertyType<Addressing, 'addressing'>,
+  hwDeviceInfo: HWDeviceInfo,
+  useUSB?: boolean = false,
+): Promise<void> => {
+  try {
+    Logger.debug('ledgerUtils::verifyAddress called')
+    Logger.debug('hwDeviceInfo', hwDeviceInfo)
+    Logger.debug('path', addressing.path)
+
+    if (hwDeviceInfo == null) {
+      throw new Error('ledgerUtils::verifyAddress: hwDeviceInfo is null')
+    }
+
+    verifyFromBip44Root(addressing)
+
+    const addressPtr = await normalizeToAddress(address)
+
+    const stakingKeyAddressing = {}
+    if (isHaskellShelley(walletImplementationId)) {
+      const baseAddr = await BaseAddress.from_address(addressPtr)
+      if (baseAddr) {
+        const rewardAddr = await RewardAddress.new(
+          Number.parseInt(CONFIG.NETWORKS.HASKELL_SHELLEY.CHAIN_NETWORK_ID, 10),
+          await baseAddr.stake_cred(),
+        )
+        const addressPayload = Buffer.from(
+          await (await rewardAddr.to_address()).to_bytes(),
+        ).toString('hex')
+        stakingKeyAddressing[addressPayload] = {
+          path: [
+            CONFIG.NUMBERS.WALLET_TYPE_PURPOSE.CIP1852,
+            CONFIG.NUMBERS.COIN_TYPES.CARDANO,
+            CONFIG.NUMBERS.ACCOUNT_INDEX + CONFIG.NUMBERS.HARD_DERIVATION_START,
```

Comment:
Just FYI we'll have to support multi-account for Ledger this year most likely, so be careful of hardcoding stuff like this